ath79: add support for GL.iNet GL-X300B

johnfzc opened the pull request:
The GL-X300B is a industrial 4G LTE router based on the Qualcomm QCA9531 SoC.

Specifications:
 - Qualcomm QCA9531 @ 650 MHz
 - 128 MB of RAM
 - 16 MB of SPI NOR FLASH
 - 2x 10/100 Mbps Ethernet
 - 2.4GHz 802.11b/g/n
 - 1x USB 2.0 (vbus driven by GPIO)
 - 4x LED, driven by GPIO
 - 1x button (reset)
 - 1x mini pci-e slot (vcc driven by GPIO)
 - RS-485 Serial Port (untested)

Flash instructions:

This firmware can be flashed using either sysupgrade from the GL.iNet
firmware or the recovery console as follows:

 - Press and hold the reset button
 - Connect power to the router, wait five seconds
 - Manually configure 192.168.1.2/24 on your computer, connect to 192.168.1.1
 - Upload the firmware image using the web interface

RS-485 serial port is untested and may depend on the following commit in
the GL.iNet repo:

https://github.com/gl-inet/openwrt/commit/202e83a32ae308fbb70502b6dbe3bb0bf8b1fba9

MAC addresses as verified by OEM firmware:

vendor   OpenWrt   address
WAN      eth0      label
LAN      eth1      label + 1
2g       phy0      label + 2

The label MAC address was found in the art partition at 0x0

Based on vendor commit:

https://github.com/gl-inet/openwrt/commit/16c5708b207eb76ff19a040dc973e560d3d8074b

Signed-off-by: John Marrett <johnf@zioncluster.ca>

johnfzc committed:
ath79: add support for GL.iNet GL-X300B

The GL-X300B is a industrial 4G LTE router based on the Qualcomm QCA9531 SoC.

Specifications:
 - Qualcomm QCA9531 @ 650 MHz
 - 128 MB of RAM
 - 16 MB of SPI NOR FLASH
 - 2x 10/100 Mbps Ethernet
 - 2.4GHz 802.11b/g/n
 - 1x USB 2.0 (vbus driven by GPIO)
 - 4x LED, driven by GPIO
 - 1x button (reset)
 - 1x mini pci-e slot (vcc driven by GPIO)
 - RS-485 Serial Port (untested)

Flash instructions:

This firmware can be flashed using either sysupgrade from the GL.iNet
firmware or the recovery console as follows:

 - Press and hold the reset button
 - Connect power to the router, wait five seconds
 - Manually configure 192.168.1.2/24 on your computer, connect to 192.168.1.1
 - Upload the firmware image using the web interface

RS-485 serial port is untested and may depend on the following commit in
the GL.iNet repo:

https://github.com/gl-inet/openwrt/commit/202e83a32ae308fbb70502b6dbe3bb0bf8b1fba9

MAC addresses as verified by OEM firmware:

vendor   OpenWrt   address
WAN      eth0      label
LAN      eth1      label + 1
2g       phy0      label + 2

The label MAC address was found in the art partition at 0x0

Based on vendor commit:

https://github.com/gl-inet/openwrt/commit/16c5708b207eb76ff19a040dc973e560d3d8074b

Signed-off-by: John Marrett <johnf@zioncluster.ca>

adschm commented:
> Use the uboot recovery console (hold reset while powering on for five seconds)
> to flash the firmware,

Please describe what to do exactly.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dts:
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+/dts-v1/;
```
Drop this.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
drop this line.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca953x.dtsi"
```
Move this line before the other includes.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+
+/ {
+	compatible = "glinet,gl-x300b", "qca,qca9531";
+	model = "GL.iNet GL-X300B";
```
compatible/model can be dropped in a DTSI.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+	model = "GL.iNet GL-X300B";
+
+	keys {
+		compatible = "gpio-keys-polled";
```
Use gpio-keys and drop poll-interval.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+		pinctrl-names = "default";
+		pinctrl-0 = <&jtag_disable_pins>;
+
+		button0 {
```
"reset"

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+		compatible = "gpio-leds";
+
+		wlan2g {
+			label = "gl-x300b:green:wlan2g";
```
drop model prefix, just "green:wlan2g" etc.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+	};
+    gpio-export {
```
Broken indent, missing empty line.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+		compatible = "gpio-export";
+
+		gpio_lte_power {
+			gpio-export,name = "gpio0";
```
"lte-power"?

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+		};
+
+		gpio_rs485tx_en {
+			gpio-export,name = "gpio1";
```
some useful name ...

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+		};
+
+		gpio_ble_rst {
+			gpio-export,name = "gpio16";
```
some useful name ...

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+		};
+
+    };
+    watchdog {
```
Indent, empty line ...

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+        compatible = "hw_wdt";
+        dog_en_gpio= <12>;
+        feed_dog_gpio=<2>;
+        feed_dog_interval=<100000000>;
```
missing spaces

johnfzc replied:
I believe I've resolved this but may not have done so correctly, please let me know if there's an issue.

johnfzc replied:
Ok, I now understand what you meant in the updated review, resolved, sorry for my confusion.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+        feed_dog_gpio=<2>;
+        feed_dog_interval=<100000000>;
+    };
+
```
drop empty line.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+
+};
+
+
```
drop second empty line.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+
+&uart {
+	status = "okay";
+    rs485_pin=<1>;
```
Indent and space, I will stop commenting on this now, please check the rest of the file as well.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
+};
```
Since you have no USB LED, the following is enough here:

```
&usb0 {
	status = "okay";
};
```

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+&spi {
+	status = "okay";
+	num-cs = <2>;
+	cs-gpios = <0>, <0>;
```
drop num-cs and cs-gpios.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+				reg = <0x050000 0x010000>;
+				read-only;
+			};
+		};
```
Add an inline comment about missing partition that will be found in DTS files.

adschm replied:
Just closing this without doing anything is not enough to make me forget. But it will influence my mood ...

Please put `/* Firmware partition is found in DTS */` or similar after the last node.

johnfzc replied:
I did add a comment, as you saw while continuing your review. I'm working to address your [questions](https://github.com/openwrt/openwrt/pull/4246#discussion_r654963051) about it and will answer you there.

adschm replied:
Another time, marked as resolved, no change. Since you won't comply to my remarks, I will stop reviewing this now.

johnfzc replied:
I'm sorry, I don't understand what you are talking about, I removed the nand partition after establishing that it wasn't present on the device, the details are in the next section of the commit. You complained in your last review about the missing comment you had requested, which was in fact present in the last commit.

I've used OpenWrt for years and dealt with many different members of the core team on a variety of occasions, they have always been welcoming and supportive. Your attitude could at best be described as harsh, and more accurately as vindictive and mean spirited.

The support for this device is quite important to me and I am working very hard to meet the appropriate standards for inclusion.

If you, @adschm, are unwilling to continue your review (for issues actually solved and addressed) then I hope another member of the community will step forwards in your place.

adschm replied:
I said "Please put /* Firmware partition is found in DTS */ or similar after the last node.".

Did you do that? No.

adschm replied:
And regarding my tone: Yes, I'm getting harsh at some point. On the other hand, I'm doing tons of device support review. This is really time-consuming. So, what I really hate is having to do the same thing multiple times, because that's just waste of life-time. So, if I've been asking to do a simple thing and you just haven't done it _twice_, I get annoyed.

johnfzc replied:
I'm confused and assure you that I am trying to do my best to answer your brief comments. They are no doubt more easily understood by someone much more experienced in this space than myself. I believed that this firmware partition comment request was related to the nand flash that was initially included in this commit and is now removed. I now think I understand what you are requesting and should be able to complete the required edit. It's because, exactly as you said, the firmware partition is in the dts and not the dtsi.

The challenge is that your experience allows you to rapidly review these commits and identify issues, however people that are submitting them, like myself, don't have the same knowledge and consequently don't necessarily understand your feedback.

I will correct this issue tomorrow and hope that you (or someone else) will continue you review. I will point out that if support for this device wasn't very important to me I would have abandoned this commit several revisions ago, to the detriment of anyone who wants to use it. I can understand that you are getting burned out on reviewing these device support commits but there has to be better answer than tearing apart the efforts of people trying to improve OpenWrt.

adschm replied:
I still don't understand what's so hard to grasp in "Please put /* Firmware partition is found in DTS */ or similar after the last node." ...

Anyway, what I want is this: vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

```
		nor_partitions: partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "u-boot";
				reg = <0x000000 0x040000>;
				read-only;
			};

			partition@40000 {
				label = "u-boot-env";
				reg = <0x040000 0x010000>;
			};

			art: partition@50000 {
				label = "art";
				reg = <0x050000 0x010000>;
				read-only;
			};

			/* Firmware partition is found in DTS */
		};
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

johnfzc replied:
The required edits are very clear to me now, I didn't think you would be reviewing today, I'll get them out to you in the next 2 hours.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+
+			nand_ubi: partition@0 {
+				label = "nand_ubi";
+				reg = <0x000000 0x8000000>;
```
just 0x0

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+/*
+&bootargs {
+	bootargs="";
+};*/
```
remove.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+	status = "okay";
+	mtd-mac-address = <&art 0x0>;
+	phy-handle = <&swphy4>;
+	ifname = "eth0";
```
What?

johnfzc replied:
I'm sorry, I'm not certain what the issue is here, the port is functional.

adschm replied:
ifname should be removed here (probably)

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+&eth1 {
+	mtd-mac-address = <&art 0x0>;
+	mtd-mac-address-increment = <1>;
+	ifname = "eth1";
```
What about 0x6?

johnfzc replied:
I'm sorry, I don't understand this comment.

adschm replied:
ifname should be removed.
mtd-mac-address is missing.
Is a valid address found in 0x6 that could be used instead of 0x0 +1?

karlp replied:
is there any particular reason it should be at 6?

johnfzc replied:
I believe 0x6 would be 6 bytes after 0x0, the length of a MAC address.

There is no other readable value however, the MAC is at 0x0, the [vendor OpenWrt](https://github.com/gl-inet/openwrt/commit/16c5708b207eb76ff19a040dc973e560d3d8074b#diff-a55c73a0a2b266ec19cdd347a344d9cffccaa6122ec09f7bc624147a25fbc77cR152-R156) uses the offsets in my commit and validation shows nothing useful at 0x6:

```
root@OpenWrt:~# echo $(mtd_get_mac_binary art 0x0)
94:83:c4:0c:05:ba
root@OpenWrt:~# echo $(mtd_get_mac_binary art 0x6)
00:03:7f:11:23:c6
```

adschm replied:
okay, then 0x0 +1 is fine.
@karlp On ath79, the typical assignment is the following:
lan 0x0
wan 0x6
wifi0 0x1002 (in caldata)
wifi1 0x5006 (in caldata)

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+&wmac {
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x1002>;
```
mtd-mac-address can be removed here.

adschm replied:
Count=3

johnfzc replied:
I'm sorry, I thought I had cleaned these issues up when I addressed the other status and spacing issues, my apologies.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,172 @@
+};
+
+&wmac {
+	status = "okay";
```
Add empty line after status=okay (always).

adschm replied:
Another resolve without a change. (Count=2)

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -218,6 +218,11 @@ qxwlan,e750a-v4-16m)
 	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth1"
 	;;
+glinet,gl-x300b|\
+    ucidef_set_led_netdev "wlan2g" "WLAN2G" "gl-x300b:green:wlan2g" "wlan0"
```
this is already in DTS and can be dropped here.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -218,6 +218,11 @@ qxwlan,e750a-v4-16m)
+glinet,gl-x300b|\
+    ucidef_set_led_netdev "wlan2g" "WLAN2G" "gl-x300b:green:wlan2g" "wlan0"
+    ucidef_set_led_netdev "wan"    "WAN"    "gl-x300b:green:wan"    "eth0"
+    ucidef_set_led_netdev "3gnet"  "3GNET"  "gl-x300b:green:lte"     "3g-wan"
```
Why not do it like everybody else? And broken indent ...

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1236,6 +1236,16 @@ define Device/glinet_gl-usb150
+  SOC := qca9531
+  DEVICE_VENDOR := GL.iNet
+  DEVICE_MODEL := GL-X300B
+  DEVICE_PACKAGES := kmod-usb2 block-mount
```
Why block-mount?

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1236,6 +1236,16 @@ define Device/glinet_gl-usb150
+  DEVICE_MODEL := GL-X300B
+  DEVICE_PACKAGES := kmod-usb2 block-mount
+  IMAGE_SIZE := 16000k
+  SUPPORTED_DEVICES += gl-x300b glinet,gl-x300b
```
The second one is already added automatically. Why would you need the first one?

johnfzc commented:
I believe I've addressed almost all of the issues identified, some feedback was unclear to me, probably because of my lack of experience in this space. I'm eager to address the outstanding issue. Once this is complete I will apply outstanding corrections on #4245 .

Thanks again for your help and patience, really appreciate it.

johnfzc commented:
@adschm thank you very much for taking the time to review my commit, I've made most of the the changes you have requested however I didn't understand some of the feedback you've provided. Can you please take a look at the updated code and outstanding unresolved comments and clarify them so that I can address them. Thank you again for your time.

Regarding the dtsi and dts files I used a dtsi as support for variants of this device could benefit from it. if you would prefer I can integrate the dtsi into the dts file. I do not plan to work on support myself, at present, beyond these two base images.

I've amended my commit and it now shows as a single change from a git perspective.

I have updated #4245 as well and will continue to do so based on your feedback here.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,157 @@
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+
```
Drop double empty line.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,157 @@
+	gpio-export {
+		compatible = "gpio-export";
+
+		gpio_minipcie {
```
Since these are all child nodes of gpio-export, you could remove the obsolete "gpio_" prefix. All cosmetic, though.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,157 @@
+		compatible = "hw_wdt";
+		dog_en_gpio= <12>;
+		feed_dog_gpio=<2>;
+		feed_dog_interval=<100000000>;
```
Please make spaces around "=" nice ...

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,157 @@
+&uart {
+	status = "okay";
+	rs485_pin=<1>;
+};
```
status = okay is already present in parent DTS.
I don't know what the other option is good for. If you need it, please fix spaces around = ...

johnfzc replied:
I don't know if the setting is necessary, I don't use the RS485 interface and can't test it so I will leave it in, cleanly formatted.

adschm replied:
Well, if you cannot resolve it, then please don't hide it by marking it as resolved.

johnfzc replied:
It wasn't my intention to hide it; it's resolved to the best of my abilities, how should we proceed from here, should it be left in but untested, or should I remove the rs485 pin configuration?

adschm replied:
Well, it's not resolved. And if we close it, nobody will see the open issue and have the chance to respond and enlighten us.
So, leaving things that are not resolved open seems a good strategy to getting additional input to me ...

BTW, `grep -rn "rs..._pin" target/linux/` does not return anything ...

johnfzc replied:
Ok, this is [patch](https://github.com/gl-inet/openwrt/blob/openwrt-19.07.7/target/linux/ath79/patches-4.14/821-fix-glinet-rs485-auto-txrx.patch#L75-L78) is present in the GL.iNet source and makes use of the `rs485_pin` setting.

I don't intend to bring this patch over; I'll remove the setting. I'm not certain if the uart is still needed; other dts files for devices with functional serial ports don't include a uart section, I believe because of your [commit](https://github.com/openwrt/openwrt/commit/3a4b751110b32380cf827f14a203190cfbc2e40f) to centralize it in the `ar9330.dtsi` file.

Please confirm if my understanding is correct and I'll remove the entire section.

Thanks for again for your time.

johnfzc replied:
I've removed the uart section and the rs485 pin, I've also updated the commit message to reflect the untested nature of the RS485 support and possible need for the patch referenced above.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,157 @@
+	};
+
+    // This flash may be present in the device but isn't
+    // currently supported in OpenWrt
```
Broken indent.

"May" means it's sometimes there and sometimes isn't? What happens if this node is present and the flash is not there?

johnfzc replied:
After investigation the flash is not present on any model, this was in the upstream commits, probably taken from the related GL-XE300 which does have this nand flash. It's removed.

adschm commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,157 @@
+			};
+		};
+	};
+
```
drop empty line.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -218,6 +218,10 @@ qxwlan,e750a-v4-16m)
 	ucidef_set_led_switch "lan" "LAN" "green:lan" "switch0" "0x02"
 	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth1"
 	;;
+glinet,gl-x300b|\
```
So, not tested.

johnfzc replied:
I did test the firmware and noticed some issues with the lights, with the other changes they are now resolved.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -218,6 +218,10 @@ qxwlan,e750a-v4-16m)
 	;;
+glinet,gl-x300b|\
+	ucidef_set_led_netdev "wan"   "WAN"   "green:wan" "eth0"
+	ucidef_set_led_netdev "3gnet" "3GNET" "green:lte" "3g-wan"
```
I don't think this interface exists ...

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -218,6 +218,10 @@ qxwlan,e750a-v4-16m)
 	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth1"
 	;;
+glinet,gl-x300b|\
+	ucidef_set_led_netdev "wan"   "WAN"   "green:wan" "eth0"
```
Either this is wrong or your setup in 02_network is.

johnfzc replied:
It was wrong, it's been corrected.

adschm commented:
Please add a MAC address overview to your commit message:
https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

johnfzc commented:
I believe that every issue identified in the review is now addressed, I have updated the commit message to include the MAC table and amended the installation process. It is possible to switch to OpenWrt using sysupgrade on this platform.

Please let me know if there are any remaining issues @adschm, thank you for your time.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -218,6 +218,9 @@ qxwlan,e750a-v4-16m)
 	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth1"
 	;;
+glinet,gl-x300b)
+	ucidef_set_led_netdev "wan"   "WAN"   "green:wan" "eth1"
```
Hmm, how could you improve this one ....?

johnfzc replied:
I thought that I had validated and cleaned up the alignment after removing the other entries, I see that I failed. I will correct this shortly.

adschm replied:
Well, even better would be to merge this case with another one that has the same content. You won't have to search for long ...

johnfzc replied:
Indeed, the required and corrected entry was only a line away.

johnfzc commented:
I believe that all outstanding issues are fully addressed, I'm confident that the fix I've used for the rs485_pin is the right solution. Can you please reconsider your previous position and complete the review @adschm. Despite the challenges we both faced I assure you my intentions are to resolve all of the outstanding issues and get this device into OpenWrt, I'm far from the only person [interested](https://forum.openwrt.org/t/failed-sysupgrade-with-invalid-tar-on-gl-xe300/98144/14?u=johnfzc) in it.

blocktrron commented on target/linux/ath79/dts/qca9531_glinet_gl-x300b.dtsi:
```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
Move everything from the dtsi to the dts, as i don't see another consumer.

johnfzc replied:
There are two [different versions](https://www.gl-inet.com/products/gl-x300b/) (scroll down to model comparison) of this router, the -RS485 and the -IOT. The IOT version includes a Bluetooth or Zigbee module. This PR only supports the hardware in the RS485, it should work on the -IOT but without support for the module. If someone were to develop support for the IOT version they would be able to use the DTSI.

Should I merge them as you requested, or leave them separate allowing someone to implement support for the IOT version more easily in the future?

johnfzc replied:
As discussed on IRC I will merge the dts and dtsi files.

blocktrron commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -218,6 +218,7 @@ qxwlan,e750a-v4-16m)
 	ucidef_set_led_switch "lan" "LAN" "green:lan" "switch0" "0x02"
 	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth1"
 	;;
+glinet,gl-x300|\
```
```
compatible = "glinet,gl-x300b"
```

So this looks wrong?

johnfzc replied:
Sorry about that mistake, resolved.

blocktrron reviewed (commented):
Please also add the switch config to `02_network`, even if only CPU + 1 port is used.

johnfzc commented:
> Please also add the switch config to `02_network`, even if only CPU + 1 port is used.

I'm not certain how to add this device to `02_network`. It appears to be getting some kind of default config that includes a WAN and LAN port, the interfaces are correctly enumerated with this config. I looked to see how this was handled with a similar device, the gl-mifi, but I don't see it or anything else I could use as a reference.

blocktrron commented:
@johnfzc 

Can you provide the output of `swconfig dev switch0 show`?

johnfzc commented:
Here's the swconfig output:

```
Global attributes:
        enable_vlan: 0
        ar8xxx_mib_poll_interval: 500
        ar8xxx_mib_type: 0
        enable_mirror_rx: 0
        enable_mirror_tx: 0
        mirror_monitor_port: 0
        mirror_source_port: 0
        arl_table: address resolution table
Port 0: MAC 94:83:c4:0c:05:bb
Port 4: MAC 00:e0:4c:bd:26:38

Port 0:
        mib: MIB counters
RxGoodByte  : 890039 (869.1 KiB)
TxByte      : 262963 (256.7 KiB)

        pvid: 0
        link: port:0 link:up speed:1000baseT full-duplex txflow rxflow
Port 1:
        mib: No MIB data
        pvid: 0
        link: port:1 link:down
Port 2:
        mib: No MIB data
        pvid: 0
        link: port:2 link:down
Port 3:
        mib: No MIB data
        pvid: 0
        link: port:3 link:down
Port 4:
        mib: MIB counters
RxGoodByte  : 263873 (257.6 KiB)
TxByte      : 124257 (121.3 KiB)

        pvid: 0
        link: port:4 link:up speed:100baseT full-duplex auto
```

From a networking perspective the WAN is eth1 and functions correctly, LAN is a bridge on eth0. I believe that eth0 goes through the switch chip to the "LAN" port on the router. I'm not certain exactly how eth1 is working and if it makes use of the switch chip; we don't see a second port active here and there's no VLAN associated with eth1, at least that I see.

blocktrron commented:
See the QCA9531 datasheet on how the switch is connected.

I was asking about the Switch config, as this allows UCI to display link state information. port 4 is the external one while port 0 is connected to the CPU MAC. Can you add the appropriate config section?

johnfzc commented:
I believe I added the appropriate configuration to `target/linux/ath79/generic/base-files/etc/board.d/02_network`, however it doesn't appear to work.

I added:

```
	glinet,gl-x300b)
		ucidef_set_interface_wan "eth1"
		ucidef_add_switch "switch0" \
			"0@eth0" "4:lan"
		;;
```

However, when I ssh through the WAN port with the LAN disconnected and boot the device I see the following log messages:

```
Fri Jul 30 15:54:45 2021 daemon.notice netifd: bridge 'br-lan' link is up
Fri Jul 30 15:54:45 2021 daemon.notice netifd: Interface 'lan' has link connectivity
Fri Jul 30 15:54:45 2021 daemon.notice netifd: Network device 'eth0' link is up
Fri Jul 30 15:54:45 2021 daemon.notice netifd: VLAN 'eth0.1' link is up
```

This occurs even though port 4 is down:

```
root@OpenWrt:~# swconfig dev switch0 show
[...]
Port 4:
        mib: No MIB data
        pvid: 1
        link: port:4 link:down
```

Can you please tell me what I did wrong either with the change or validating it? Thanks very much for your help!

blocktrron commented:
When installing LuCI, can you see the port state at Network --> switch?

johnfzc commented:
Yes, the port state in the luci interface is working correctly.

blocktrron commented:
Merged to my staging tree, thanks!
